azure-core/azure/core/pipeline/transport/requests_basic.py

```diff
@@ -0,0 +1,205 @@
+from azure.core.exceptions import (
+    ServiceRequestError,
+    ConnectionError,
+    ConnectionReadError,
+    ReadTimeoutError,
+    raise_with_traceback
+)
+
+# Matching requests, because why not?
+CONTENT_CHUNK_SIZE = 10 * 1024
+
+class RequestsContext(object):
+    def __init__(self, session, transport):
+        self.session = session
+        self.transport = transport
+
+
+class _RequestsTransportResponseBase(_HttpResponseBase):
+
+    def __init__(self, request, requests_response):
+        super(_RequestsTransportResponseBase, self).__init__(request, requests_response)
+        self.status_code = requests_response.status_code
+        self.headers = requests_response.headers
+        self.reason = requests_response.reason
+
+    def body(self):
+        return self.internal_response.content
+
+    def text(self, encoding=None):
+        if encoding:
+            self.internal_response.encoding = encoding
+        return self.internal_response.text
+
+    def raise_for_status(self):
+        self.internal_response.raise_for_status()
+
+
+class RequestsTransportResponse(_RequestsTransportResponseBase, HttpResponse):
+
+    def stream_download(self, chunk_size=None, callback=None):
+        # type: (Optional[int], Optional[Callable]) -> Iterator[bytes]
+        """Generator for streaming request body data.
+
+        :param callback: Custom callback for monitoring progress.
+        :param int chunk_size:
+        """
+        chunk_size = chunk_size or CONTENT_CHUNK_SIZE
+        with contextlib.closing(self.internal_response) as response:
+            # https://github.com/PyCQA/pylint/issues/1437
+            for chunk in response.iter_content(chunk_size):  # pylint: disable=no-member
+                if not chunk:
+                    break
+                if callback and callable(callback):
+                    callback(chunk, response=response)
+                yield chunk
+
+
+class RequestsTransport(HttpTransport):
+    """Implements a basic requests HTTP sender.
+
+    Since requests team recommends to use one session per requests, you should
+    not consider this class as thread-safe, since it will use one Session
+    per instance.
+
+    In this simple implementation:
+    - You provide the configured session if you want to, or a basic session is created.
+    - All kwargs received by "send" are sent to session.request directly
+    """
+
+    _protocols = ['http://', 'https://']
+
+    def __init__(self, configuration=None, session=None):
+        # type: (Optional[requests.Session]) -> None
+        self._session_mapping = threading.local()
+        self.config = configuration or Configuration()
+        self.session = session or requests.Session()
+
+    def __enter__(self):
+        # type: () -> RequestsTransport
+        return self
+
+    def __exit__(self, *exc_details):  # pylint: disable=arguments-differ
+        self.close()
+
+    def _init_session(self, session):
```

Comment:
Should we do this if a session is passed in to the transport? Or only if a session is conjured up by the client/transport?

Reply:
I think we should do it for both a new session and a passed in session. Otherwise a user might be confused if they configured proxies and wondered why they weren't being applied to the session....
We also need to do it to disable retries in a passed in session.

Reply:
But if I pass in the same session to two different clients, or I have configured my session to my liking (which is presumably two of the reasons that I'd provide a session in the first place), does that make sense?

Reply:
Probably true - but referring to my above comment.... do we have any concrete scenarios to support passing in a session anyway?
If a session is already kept alive to reuse the connection pool when needed - what other reasons would a user or developer need to create/customize the session that they couldn't do with a policy (which exposes the session on the context).